### DISCUSSION TITLE
[Feature] Direct link in shares

### DISCUSSION THREAD

**Discussion by lordkitsuna** [immich-app/immich#3556] [Feature Request]
### The feature

When doing a single image or video share add an option to link directly to the file instead of just a full immich page as an option. This will allow for embedding of video or images in things like discord. 

as an example of what i mean seafile does this for filesharing already, in this example link, there is the "in a web page" like what immich does right now 

https://seafile.kitsuna.net/f/99b0b7d057bd4051a060/

however it _also_ has the option to give me a raw link, which will allow things like discord to embedded the video directly

https://seafile.kitsuna.net/f/99b0b7d057bd4051a060/?dl=1

at the moment if you share a video with immich you just get a static picture as the preview 

![image](https://github.com/immich-app/immich/assets/3444267/b329a15d-4e23-4ff0-be1a-e1e7851fc971)

### Platform

- [X] Server
- [X] Web
- [ ] Mobile

**Comment by trevor-media** [comment]
I also really need this feature instead of AWS S3, looking forward to your guidance.

**Comment by mptpro** [comment]
I too would like this feature - it is the one missing thing for me to make Immich perferct!

**Comment by bo0tzz** [comment]
When you open the shared link viewer in a browser, you can right-click the image and hit "Copy Image Link" to get a direct link that'll embed in Discord and such.

![image](https://github.com/user-attachments/assets/10318c8c-2cd9-4a36-9017-f1f574679e97)

**Reply by lordkitsuna** [reply]
> When you open the shared link viewer in a browser, you can right-click the image and hit "Copy Image Link" to get a direct link that'll embed in Discord and such.
> 
> ![image](https://private-user-images.githubusercontent.com/11602424/369135448-10318c8c-2cd9-4a36-9017-f1f574679e97.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjY3NzUyOTMsIm5iZiI6MTcyNjc3NDk5MywicGF0aCI6Ii8xMTYwMjQyNC8zNjkxMzU0NDgtMTAzMThjOGMtMmNkOS00YTM2LTkwMTctZjFmNTc0Njc5ZTk3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTE5VDE5NDMxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThlY2E3MzE4NGYzMTQwODBiNGVhZDVlMzE2ZTQ1ODJjYjg4YWNlNzRhOTg2OTAyNjMyZGYwZmQ5ZmM4MDMwMTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tvBFNjXCcrCfqT4EWFpMsOKur0oxmeS9FYvGFArobVg)

This does not appear to work for video. I get a link but it fails to embed. And still drags the user to the full immich page if put in the browser. I'd be nice to see a proper direct file link provided by immich.

As I showed in the original example both a link with a surrounding web app and a direct link to the file with no surrounding web page (browsers will simply download the file without opening any page on direct links) are possible and that's what I'm looking for here

**Reply by TheColorman** [reply]
Choosing "Copy video address" in chrome gives me a link that points to `/api/assets/...` that does in fact embed and play just fine in discord.
![image](https://github.com/user-attachments/assets/d6a8dab3-cbc1-409a-a997-515b86e591a7)

**Comment by danieldietzler** [comment]
While a good workaround, this probably isn't perfect and doesn't work (well) for mobile either.

**Comment by yourfate** [comment]
I would also really like this feature, just have a "direct link to original" option in the share sheet.

with the workaround from @bo0tzz I don't get the full resolution file unless I zoom in, too.

Also, that link is suuuper long and has no file extension at the end.

**Comment by nsrafk** [comment]
Agreed, this would be a nice addition. Very useful when only sharing a single file

**Comment by zhdenny** [comment]
I would like to see this feature as well. Being able to embed your shared image link directly into chats/forums/etc is very useful

**Comment by lucasantarella** [comment]
Another consideration would be to use Signed URLs for timed expiration and further controls. We can probably re-use the existing AWS Cloudfront signature method of pub/priv key signature on the URL. This would allow url sharing without signing in, similar to Google Photos, and would be pretter secure.

I can work on the backend implementation of this but would need some assistance with the frontend and mobile.

**Reply by bo0tzz** [reply]
I don't think it's necessary to create a new way of getting the file - as illustrated above, that's already possible. It just needs a bit of polishing and a way to get the link from the UI.

**Reply by lucasantarella** [reply]
Maybe I'm missing it tho, are the urls that can be copied and pasted already secured? Do they require authentication to open? If not, maybe this is already a security hole.

Ideally the urls that would be shared would be signed so a malicious actor couldn't brute force guess other paths.

**Reply by danieldietzler** [reply]
> so a malicious actor couldn't brute force guess other paths

GLHF brute forcing that. Way too many combinations, that's the entire point of that. Just like with any API key/password

**Reply by TheColorman** [reply]
I would assume only assets you have publicly shared would be accessible directly by link, but can't test right now.

**Reply by bo0tzz** [reply]
> are the urls that can be copied and pasted already secured? Do they require authentication to open? 

Yes, they have the same authentication on them as the current shared link as a whole.

**Comment by fain126** [comment]
I know I can right-click on the shared image page to get the raw image link, but this workflow is cumbersome. The experience is especially poor on mobile devices.
I hope to add the ability to fetch raw direct links inside the share link management panel. Users should be able to get the raw image URL with one click, removing the extra steps of opening the shared page and right-clicking to copy the link.
Currently, manually opening the shared page and right-clicking is the only workaround, which is inconvenient.
